##Data Preparation


Unduh data CSV dari Github https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

Import Librari yang dibutuhkan

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("/content/BL-Flickr-Images-Book.csv")

In [3]:
print(df.dtypes)

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object


In [5]:
df.describe()

Identifier  Corporate Author  Corporate Contributors  Engraver
count  8.287000e+03               0.0                     0.0       0.0
mean   2.017344e+06               NaN                     NaN       NaN
std    1.190379e+06               NaN                     NaN       NaN
min    2.060000e+02               NaN                     NaN       NaN
25%    9.157875e+05               NaN                     NaN       NaN
50%    2.043707e+06               NaN                     NaN       NaN
75%    3.047430e+06               NaN                     NaN       NaN
max    4.160339e+06               NaN                     NaN       NaN

In [6]:
df.head()

Identifier  ...                           Shelfmarks
0         206  ...    British Library HMNTS 12641.b.30.
1         216  ...    British Library HMNTS 12626.cc.2.
2         218  ...    British Library HMNTS 12625.dd.1.
3         472  ...  British Library HMNTS 10369.bbb.15.
4         480  ...     British Library HMNTS 9007.d.28.

[5 rows x 15 columns]

In [7]:
yg_dihapus = ['Edition Statement', 'Corporate Author', 'Corporate Contributors', 'Former owner', 'Engraver', 'Shelfmarks']
df.drop(yg_dihapus, inplace=True, axis=1)

In [8]:
df.head()

Identifier  ...                                         Flickr URL
0         206  ...  http://www.flickr.com/photos/britishlibrary/ta...
1         216  ...  http://www.flickr.com/photos/britishlibrary/ta...
2         218  ...  http://www.flickr.com/photos/britishlibrary/ta...
3         472  ...  http://www.flickr.com/photos/britishlibrary/ta...
4         480  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 9 columns]

Mengubah Indeks Data

In [10]:
df['Identifier'].is_unique

True

In [11]:
df = df.set_index('Identifier')

In [12]:
df.head()

Place of Publication  ...                                         Flickr URL
Identifier                            ...                                                   
206                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
216         London; Virtue & Yorston  ...  http://www.flickr.com/photos/britishlibrary/ta...
218                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
472                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
480                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 8 columns]

In [15]:
df.loc[480] #mencari data dengan mudah berdasarkan Index yang telah dibuat

Place of Publication                                               London
Date of Publication                                                  1857
Publisher                                            Wertheim & Macintosh
Title                   [The World in which I live, and my place in it...
Author                                                          A., E. S.
Contributors                                          BROOME, John Henry.
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 480, dtype: object

Merapikan Fields

In [16]:
df.dtypes.value_counts()

object    8
dtype: int64

In [19]:
df.loc[1905:,'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

1. Hilangkan tanggal lain dalam kurung siku
2. Konversi rentang tanggal
3. Hilangkan tanggal yang tidak jelas [1897?] --> NaN
4. Konversi NaN

In [20]:
regex = r'^(\d{4})'

In [22]:
#Untuk merapikan atau ekstarak
ekstrak = df['Date of Publication'].str.extract('^(\d{4})', expand = False)
ekstrak.head(15)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
1280    1802
1808    1859
1905    1888
1929    1839
2836    1897
Name: Date of Publication, dtype: object

In [25]:
df['Date of Publication'] = pd.to_numeric(ekstrak)
df['Date of Publication'].dtype

dtype('float64')

Membersihkan Kolom untuk membuang kolom yang bisa dibuang

In [28]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

In [29]:
df.loc[667]

Place of Publication                  pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Coming of Spring, and other poems. By J. A...
Author                                                      A., J.|A., J.
Contributors                                ANDREWS, J. - Writer of Verse
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 667, dtype: object

In [33]:
df.loc[4158128]

Place of Publication                                                Derby
Date of Publication                                                1831.0
Publisher                                                 M. Mozley & Son
Title                   The History and Gazetteer of the County of Der...
Author                                         GLOVER, Stephen - of Derby
Contributors                                               NOBLE, Thomas.
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4158128, dtype: object

In [32]:
df['Place of Publication'].tail(10)

Identifier
4117751               New York
4117752                 London
4156359               New York
4157746                 London
4157862    Newcastle-upon-Tyne
4158088                 London
4158128                  Derby
4159563                 London
4159587    Newcastle upon Tyne
4160339                 London
Name: Place of Publication, dtype: object

Merapikan -->Newcastle-upon-Tyne, Newcastle upon Tyne

In [34]:
publikasi = df['Place of Publication']
london = publikasi.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [35]:
oxford = publikasi.str.contains('Oxford')

In [36]:
df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               publikasi.str.replace('-', ' ')))

In [37]:
df['Place of Publication'].head(10)

Identifier
206     London
216     London
218     London
472     London
480     London
481     London
519     London
667     Oxford
874     London
1143    London
Name: Place of Publication, dtype: object

In [38]:
df['Place of Publication'].tail(10)

Identifier
4117751               New York
4117752                 London
4156359               New York
4157746                 London
4157862    Newcastle upon Tyne
4158088                 London
4158128                  Derby
4159563                 London
4159587    Newcastle upon Tyne
4160339                 London
Name: Place of Publication, dtype: object

In [39]:
df.head()

Place of Publication  ...                                         Flickr URL
Identifier                       ...                                                   
206                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
216                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
218                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
472                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...
480                      London  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 8 columns]

Membersihkan Database dengan ApplyMap dengan data yang di unduh melalui https://raw.githubusercontent.com/realpython/python-data-cleaning/master/Datasets/university_towns.txt

In [44]:
universuty_town = []
with open ("/content/university_towns.txt") as file:
  for line in file:
    if '[edit]' in line:
      state = line
    else:
      universuty_town.append((state, line))
universuty_town[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [45]:
df_kota = pd.DataFrame(universuty_town, columns= ['State', 'RegionName'])
df_kota.head(15)

State                                         RegionName
0    Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1    Alabama[edit]\n           Florence (University of North Alabama)\n
2    Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3    Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4    Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n
5    Alabama[edit]\n                        Troy (Troy University)[2]\n
6    Alabama[edit]\n  Tuscaloosa (University of Alabama, Stillman Co...
7    Alabama[edit]\n                Tuskegee (Tuskegee University)[5]\n
8     Alaska[edit]\n    Fairbanks (University of Alaska Fairbanks)[2]\n
9    Arizona[edit]\n       Flagstaff (Northern Arizona University)[6]\n
10   Arizona[edit]\n                 Tempe (Arizona State University)\n
11   Arizona[edit]\n                   Tucson (University of Arizona)\n
12  Arkansas[edit]\n  Arkadelphia (Henderson State University, Ouach...
13  Arkansas[edit]\n  Conway (Central Baptist College, Hendrix Colle...
14  Arkansas[edit]\n         Fayetteville (University of Arkansas)[7]\n

In [46]:
contoh_data = {0 : ['aock', 'python', 'real', 'numpy'],
               1 : ['Dataset', 'Pandas', 'Python', 'Clean']
               }
contoh_dataframe = pd.DataFrame(contoh_data)
contoh_dataframe.head()

0        1
0    aock  Dataset
1  python   Pandas
2    real   Python
3   numpy    Clean

In [47]:
def get_citystate(item):
    if '(' in item:
      return item[:item.find('(')]
    elif '[' in item:
      return item[:item.find('[')]
    else:
      return item

In [48]:
df_kota = df_kota.applymap(get_citystate)

In [49]:
df_kota.head(15)

State     RegionName
0    Alabama        Auburn 
1    Alabama      Florence 
2    Alabama  Jacksonville 
3    Alabama    Livingston 
4    Alabama    Montevallo 
5    Alabama          Troy 
6    Alabama    Tuscaloosa 
7    Alabama      Tuskegee 
8     Alaska     Fairbanks 
9    Arizona     Flagstaff 
10   Arizona         Tempe 
11   Arizona        Tucson 
12  Arkansas   Arkadelphia 
13  Arkansas        Conway 
14  Arkansas  Fayetteville